### PR TITLE
Move Teleport Connect over to Vite

### PR DESCRIPTION
**Pull request by ryanclark**
This moves Connect's build & development tooling over to Vite.

In order to use Vite (which doesn't like CommonJS includes unless it's a 3rd party module), I had to swap over our JS protobuf generation to TypeScript - this is actually a nice advantage, because we can now use the protobuf types in the Teleport codebase. It also fixes a few type issues/FIXMEs in the codebase. I moved user preferences over to using the protobuf files as there was intertwined logic between Teleterm, Teleport & the gRPC representation, so I simplified it.

Sorry for the large PR. These changes were a result of changing one thing and then fixing the errors, changing another, fixing those, etc - it was hard to keep these concise and in separate, ordered commits.

### Speed improvements

Start -
Vite - 5.89 seconds
Webpack 22.7 seconds

Build -
Vite - 14.28 seconds
Vite (bytecode) - 168.26 seconds
Webpack - 92.64 seconds

### Caveats

The `oneOf` representation in `@protobuf-ts/plugin` only works if `strictNullChecks` is enabled. I tried to enable this, but I couldn't get Connect to only be checked and not Teleport.

To get around this, I added some helpers to ensure type safety. There's only 3 different oneOf values to check in the codebase, and I don't think it's too tedious to add new helpers for new values. If the protobuf representation changes, the types will break, so they will always be in sync.

### Things to test

- [x] Bundling and packaging
  - [x] `agentCleanupDaemon.js` requires `winston`, but it's not included in `node_modules`. Webpack used to bundle it into `agentCleanupDaemon.js` itself.
- [x] Passwordless login
- [x] Shell test plan
  - [x] Detecting `$HOME` doesn't seem to work, opening a new local shell tab while `DocumentCluster` is active opens `/` instead of `~`.
- [x] File transfer
- [x] [Hot module replacement](https://electron-vite.org/guide/hmr)
- [x] F6 to reload main in dev mode
  -  electron-vite supports [hot reloading of the main process](https://electron-vite.org/guide/hot-reloading), but they say "Because the timing of reloading is uncontrollable, hot reloading cannot be perfectly implemented. In practice, hot reloading is not always beneficial，so it is recommended to use the CLI option to flexibly decide whether to enable or not."
- [x] Removing unreachable code (`process.env.NODE_ENV`)
- [x] CSP matches that generated by Webpack
- `yarn dev-term` and packaged app are considered to be the same app by Electron, preventing both from running at the same time.
  - Not a blocker, we can deal with this later. Most likely a result of the dev version of the app using the same package.json (and thus package name) as the packaged prod version.
- [x] Usage events

### CODE REVIEW

**Comment by ryanclark** [comment]
Note - I had tried to add `strictNullChecks`, so there may be code changes to account for that, but they're not needed now. Please point out any you see and I'll revert

File: web/packages/teleterm/src/services/tshd/createClient.ts
Comment (ravicious) [inline web/packages/teleterm/src/services/tshd/createClient.ts:96]: 😌

File: web/packages/teleterm/src/services/tshd/createClient.ts
Comment (ravicious) [inline web/packages/teleterm/src/services/tshd/createClient.ts:914]: Yesss, now comments from protos get generated into JSDocs. Try looking up `nextKey` here.

File: web/packages/teleterm/src/services/tshd/createClient.ts
Comment (ravicious) [inline web/packages/teleterm/src/services/tshd/createClient.ts:982]: That seems like a leftover from strict null checks. This check here probably doesn't do anything bad, but I guess I'd remove it just to keep the implementation as close to the previous one as possible?

File: web/packages/teleterm/src/services/tshd/createClient.ts
Comment (ravicious) [inline web/packages/teleterm/src/services/tshd/createClient.ts:1013]: I think we can pass `userPreferences` directly.

File: build.assets/Dockerfile
Comment (ravicious) [inline build.assets/Dockerfile:295]: I'm not sure if we won't have to do this in three steps, where we first add `@protobuf-ts/plugin` to the Dockerfile, then add code that uses it and then remove `grpc-tools` and `grpc_tools_node_protoc_ts`.

I'm not up to date how the GitHub Actions work nowadays, but with GCP I remember issues with the fact that the builders were using cached/published version of the Dockerfile and not the one from the branch.

The lint action failed on this:

```
buf generate --template=buf-ts.gen.yaml --path=proto/prehog/ --path=proto/teleport/lib/teleterm/ --path=api/proto/teleport/userpreferences/ --path=api/proto/teleport/accesslist/ --path=api/proto/teleport/trait/ --path=api/proto/teleport/header/
Failure: plugin typescript: exec: "/usr/local/lib/node_modules/@protobuf-ts/plugin/bin/protoc-gen-ts": stat /usr/local/lib/node_modules/@protobuf-ts/plugin/bin/protoc-gen-ts: no such file or directory
```

https://github.com/gravitational/teleport/actions/runs/7585240077/job/20660640230?pr=36921#step:15:17

But I'm not 100% sure if it was caused by what I'm talking about.

**Review by ravicious — COMMENTED** [review]
I went through the tshd client. I'm yet to build the app, but I definitely want to look into how the final packaged app looks like structure-wise. I can also do a local signed build before we take it to CI.

File: web/packages/teleterm/src/services/tshd/createClient.ts
Comment (ryanclark) [reply web/packages/teleterm/src/services/tshd/createClient.ts]: Yep, I've removed them

File: web/packages/teleterm/src/services/tshd/createClient.ts
Comment (ryanclark) [reply web/packages/teleterm/src/services/tshd/createClient.ts]: > I'd remove it just to keep the implementation as close to the previous one as possible?

Agreed, undone

File: web/packages/teleterm/src/services/tshd/createClient.ts
Comment (ryanclark) [reply web/packages/teleterm/src/services/tshd/createClient.ts]: Yep!

File: build.assets/Dockerfile
Comment (ryanclark) [reply build.assets/Dockerfile:295]: Yeah I think you're correct. Let me split out this part into a new PR

Comment (ryanclark) [reply build.assets/Dockerfile:295]: https://github.com/gravitational/teleport/pull/37009

File: web/packages/teleterm/electron.vite.config.mts
Comment (gzdunek) [inline web/packages/teleterm/electron.vite.config.mts:156]: I don't see any warning related to `eval` if I remove this. I guess we don't need it.

Comment (ryanclark) [reply web/packages/teleterm/electron.vite.config.mts:156]: Ah, must've been from the other TypeScript protobuf gen library I had tried. I'll remove!

File: package.json
Comment (ravicious) [reply package.json]: Initially we added native deps to start-term after the hackathon because Krzysztof and Edoardo had a ton of problems. They removed the whole node_modules each time they ran into some kind of a problem with JS and then they didn't remember about reinstalling native deps. ;)

So we just put native deps there. Grzegorz knew that it'll be slow as hell on Windows, but I was like "We'll deal with that later".

Well, last week I had to finally build the app on Windows and it was slow as hell. So it's good that you added it there because I'd probably end up doing the same in a week or two.

File: web/packages/teleterm/src/ui/services/workspacesService/workspacesService.ts
Comment (gzdunek) [inline web/packages/teleterm/src/ui/services/workspacesService/workspacesService.ts:1]: Most of the changes here can be reverted here, right? I think the are leftovers from strict null checks.

File: web/packages/teleterm/src/helpers.ts
Comment (gzdunek) [inline web/packages/teleterm/src/helpers.ts:39]: This can be removed (altogether with the comment from line 19).

**Comment by ravicious** [comment]
Grzegorz and I want to go through Connect's test plan again once we merge this into master and make a dev build off of v15 with these changes. But what about the changes in Web UI, how are we going to test them?

**Comment by ravicious** [comment]
If we merge it to master and backport to v15 only after a few days, it'll be inconvenient to anyone working on the Web UI code that interacts with types from protos. Once we get everything working in this PR, we should probably do tests on the v15 dev build and only then merge to master, followed by a quick backport to v15.

File: web/packages/teleterm/package.json
Comment (ravicious) [inline web/packages/teleterm/package.json:14]: Regarding hot reloading of main process and preload: I just tried it and it seems to work okay. You just do `yarn start-term -w` and electron-vite automatically restarts the app whenever the main process or the preload code changes. It restarts on changes to the shared process and the agent cleanup daemon too.

Restarting the app steals focus, that's one minus.

@gzdunek Can you think of a situation where we'd want to rebuild the main process but not restart the app immediately?

I feel like the `-w` option works well enough, we could work with it for a week or two and then see what could be adjusted. Besides, who cares about hot reloading when you can just rerun `yarn start-term` in a few seconds.

Comment (gzdunek) [reply web/packages/teleterm/package.json:14]: >@gzdunek Can you think of a situation where we'd want to rebuild the main process but not restart the app immediately?

Nothing comes to my mind.

>I feel like the -w option works well enough, we could work with it for a week or two and then see what could be adjusted. Besides, who cares about hot reloading when you can just rerun yarn start-term in a few seconds.

I agree.

Comment (gzdunek) [reply web/packages/teleterm/package.json:14]: I see that reloading doesn't work for the renderer:
```
MAIN [Main] warn: Navigation to https://localhost:8080/ blocked by 'will-navigate'
```
I will add an exception in dev mode for localhost:8080.

File: .eslintignore
Comment (ravicious) [inline .eslintignore:7]: Did you mean to ignore `web/packages/teleterm/build`?

Comment (ravicious) [reply .eslintignore:7]: Eslint ignored the Webpack builds probably because of this:

https://github.com/gravitational/teleport/blob/218d608f835b8b15c2e8dca2a3542652d470ed4c/web/packages/build/.eslintrc.js#L37

And the files prepared by Webpack were in `web/packages/teleterm/build/app/dist`.

File: web/packages/teleterm/package.json
Comment (ravicious) [reply web/packages/teleterm/package.json:14]: The previous mechanism we had with F6 restarting the app was useful in situations where you modified just tsh and wanted to restart the app. There's probably a way to set up a rollup watcher in electron-vite that's going to restart the app on changes to `build/tsh`, but that's not essential for now.

File: web/packages/teleport/src/User/UserContext.tsx
Comment (ravicious) [inline web/packages/teleport/src/User/UserContext.tsx:101]: Could someone double check this change here?

File: web/packages/teleterm/src/services/tshd/types.ts
Comment (ravicious) [inline web/packages/teleterm/src/services/tshd/types.ts:134]: This change should be fine I think, @gzdunek AFAIK it was to accommodate for some weird Timestamp shenanigans in protos?

Comment (gzdunek) [reply web/packages/teleterm/src/services/tshd/types.ts:134]: I think so, I will make sure that the correct date is sent with the usage events.

File: web/packages/teleterm/src/types.ts
Comment (ravicious) [inline web/packages/teleterm/src/types.ts]: This is so much nicer than how I implemented it!

**Review by ravicious — COMMENTED** [review]
I went through the whole PR and everything appears to be okay just from looking at the code alone. We'll run the Connect test plan on 15.0.0-dev.ravicious.10 just to double check.

File: web/packages/teleterm/src/services/tshd/types.ts
Comment (gzdunek) [reply web/packages/teleterm/src/services/tshd/types.ts:134]: Except for the wrong environment, usage events are sent correctly.

**Review by ravicious — APPROVED** [review]
Okay, we ran the test plan against v15.0.0-dev.ravicious.10 and it looks everything works fine. What's next, do we wait for the protobuf PRs to land in master?

**Review by gzdunek — APPROVED** [review]

**Comment by ryanclark** [comment]
> Okay, we ran the test plan against v15.0.0-dev.ravicious.10 and it looks everything works fine. What's next, do we wait for the protobuf PRs to land in master?

Yep

File: web/packages/teleport/src/services/userPreferences/userPreferences.ts
Comment (gzdunek) [inline web/packages/teleport/src/services/userPreferences/userPreferences.ts:143]: I'd spread `clusterPreferences` too:
```
    clusterPreferences: {
      ...preferences.clusterPreferences,
      pinnedResources:
        preferences.clusterPreferences?.pinnedResources?.resourceIds ?? [],
    }
```
If someone adds a new property, we won't skip it. The same for `convertBackendUserPreferences`.

**Comment by gzdunek** [comment]
Something is still wrong with leaf cluster preferences, I get JS errors. Let me look into it.

File: web/packages/teleterm/src/ui/AppInitializer/AppInitializer.tsx
Comment (gzdunek) [inline web/packages/teleterm/src/ui/AppInitializer/AppInitializer.tsx:38]: `console.log`s to remove.

**Review by gzdunek — APPROVED** [review]

**Pull request merged**
